From 3c6065aec449fae023bff9b0db26677c370a6b51 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:19 -0400
Subject: [usb] Check if port status is equal to RxDetect

Message-id: <1410887463-20674-548-git-send-email-dzickus@redhat.com>
Patchwork-id: 95630
O-Subject: [RHEL7 PATCH 547/591] usb: Check if port status is equal to RxDetect
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit bb86cf569bbd7ad4dce581a37c7fbd748057e9dc
Author: Gavin Guo <gavin.guo@canonical.com>
Date:   Fri Jul 18 01:12:13 2014 +0800

    usb: Check if port status is equal to RxDetect

    When using USB 3.0 pen drive with the [AMD] FCH USB XHCI Controller
    [1022:7814], the second hotplugging will experience the USB 3.0 pen
    drive is recognized as high-speed device. After bisecting the kernel,
    I found the commit number 41e7e056cdc662f704fa9262e5c6e213b4ab45dd
    (USB: Allow USB 3.0 ports to be disabled.) causes the bug. After doing
    some experiments, the bug can be fixed by avoiding executing the function
    hub_usb3_port_disable(). Because the port status with [AMD] FCH USB
    XHCI Controlleris [1022:7814] is already in RxDetect
    (I tried printing out the port status before setting to Disabled state),
    it's reasonable to check the port status before really executing
    hub_usb3_port_disable().

    Fixes: 41e7e056cdc6 (USB: Allow USB 3.0 ports to be disabled.)
    Signed-off-by: Gavin Guo <gavin.guo@canonical.com>
    Acked-by: Alan Stern <stern@rowland.harvard.edu>
    Cc: <stable@vger.kernel.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 75014df..4ae2748 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -888,6 +888,25 @@ static int hub_usb3_port_disable(struct usb_hub *hub, int port1)
  if (!hub_is_superspeed(hub->hdev))
   return -EINVAL;
 
+ ret = hub_port_status(hub, port1, &portstatus, &portchange);
+ if (ret < 0)
+  return ret;
+
+ /*
+  * USB controller Advanced Micro Devices, Inc. [AMD] FCH USB XHCI
+  * Controller [1022:7814] will have spurious result making the following
+  * usb 3.0 device hotplugging route to the 2.0 root hub and recognized
+  * as high-speed device if we set the usb 3.0 port link state to
+  * Disabled. Since it's already in USB_SS_PORT_LS_RX_DETECT state, we
+  * check the state here to avoid the bug.
+  */
+ if ((portstatus & USB_PORT_STAT_LINK_STATE) ==
+    USB_SS_PORT_LS_RX_DETECT) {
+  dev_dbg(&hub->ports[port1 - 1]->dev,
+    "Not disabling port; link state is RxDetect\n");
+  return ret;
+ }
+
  ret = hub_set_port_link_state(hub, port1, USB_SS_PORT_LS_SS_DISABLED);
  if (ret)
   return ret;
-- 
1.7.1